iot/end_to_end_example/server/src/server.kt

```diff
@@ -0,0 +1,126 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+package com.google.iot
+
+import com.google.cloud.pubsub.v1.AckReplyConsumer
+import com.google.cloud.pubsub.v1.MessageReceiver
+import com.google.cloud.pubsub.v1.Subscriber
+
+import com.google.pubsub.v1.ProjectSubscriptionName
+import com.google.pubsub.v1.PubsubMessage
+
+import com.google.cloud.iot.v1.DeviceManagerClient
+import com.google.cloud.iot.v1.DeviceName
+import com.google.cloud.iot.v1.ModifyCloudToDeviceConfigRequest
+
+import com.google.protobuf.ByteString
+
+import java.io.IOException
+import java.util.concurrent.LinkedBlockingDeque
+
+import com.beust.klaxon.Klaxon
+import com.beust.klaxon.JsonObject
+
+private val iot = DeviceManagerClient.create()
+
+data class IotData(val temperature: Int)
+
+// The main loop. Consumes messages from the Pub/Sub subscription.
+fun main(args: Array<String>) {
+    if (args.size < 2) {
+        error("Usage: java -jar server.jar PROJECT_ID SUBSCRIPTION_ID\n")
+    }
+
+    val projectId = args[0]
+    val subscriptionId = args[1]
+    val messages = LinkedBlockingDeque<PubsubMessage>()
+
+    class MessageReceiverExample : MessageReceiver {
+        override fun receiveMessage(message: PubsubMessage, consumer: AckReplyConsumer) {
+            messages.offer(message)
+            consumer.ack()
+        }
+    }
+
+    val subscriptionName = ProjectSubscriptionName.of(projectId, subscriptionId)
+
+    val subscriber = Subscriber
+        .newBuilder(subscriptionName, MessageReceiverExample()).build()
+
+    try {
+        // Create a subscriber bound to the asynchronous message receiver
+        subscriber.startAsync().awaitRunning()
+
+        // Continue to listen to messages
+        println("Listening to messages on $subscriptionId:")
+        while (true) {
+            val message = messages.take()
+            handle_pubsub_message(message)
+        }
+    } finally {
+        subscriber.stopAsync()
+    }
+}
+
+fun handle_pubsub_message(message: PubsubMessage) {
+    println("Message Id: ${message.messageId} Data: ${message.data.toStringUtf8()}")
+
+    val data: IotData? = Klaxon().parse<IotData>(message.data.toStringUtf8())
+    if (data == null) {
+        println("Loading JSON payload failed.")
+        return
+    }
+
+    // Get the registry id and device id from the attributes. These are
+    // automatically supplied by IoT, and allow the server to determine which
+    // device sent the event.
+    val deviceId = message.attributesMap["deviceId"]
+    val deviceName = DeviceName.format(
+        message.attributesMap["projectId"],
+        message.attributesMap["deviceRegistryLocation"],
+        message.attributesMap["deviceRegistryId"],
+        deviceId
+    )
+
+    println("The device ($deviceId) has a temperature of ${data.temperature}")
+
+    if (data.temperature < 0) {
+        // Turn off the fan.
+        update_device_config(deviceName, fanOn = false)
```

Comment:
the underscored method name is not idiomatic in Java/Kotlin. camel casing is better.